instrumentation/rspec/lib/opentelemetry/instrumentation/rspec/formatter.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'opentelemetry-api'
+require_relative './version'
+
+module OpenTelemetry
+  module Instrumentation
+    module RSpec
+      # An RSpec Formatter that outputs Otel spans
+      class Formatter
+        attr_reader :output
+
+        ::RSpec::Core::Formatters.register self, :example_started, :example_finished, :example_group_started, :example_group_finished, :start, :stop
+
+        @clock = Time.method(:now)
+
+        def self.current_timestamp
+          @clock.call
+        end
+
+        def initialize(output = StringIO.new, tracer_provider = OpenTelemetry.tracer_provider)
+          @spans_and_tokens = []
+          @output = output
+          @tracer_provider = tracer_provider
+        end
+
+        def tracer
+          @tracer ||= @tracer_provider.tracer('OpenTelemetry::Instrumentation::RSpec', OpenTelemetry::Instrumentation::RSpec::VERSION)
+        end
+
+        def current_timestamp
+          self.class.current_timestamp
+        end
+
+        def start(notification)
+          span = tracer.start_span('RSpec suite', start_timestamp: current_timestamp)
+          track_span(span)
+        end
+
+        def stop(notification)
+          pop_and_finalize_span
+        end
+
+        def example_group_started(notification)
+          description = notification.group.description
+          span = tracer.start_span(description, start_timestamp: current_timestamp)
+          track_span(span)
+        end
+
+        def example_group_finished(notification)
+          pop_and_finalize_span
+        end
+
+        def example_started(notification)
+          example = notification.example
+          attributes = {
+            'location' => example.location.to_s,
+            'full_description' => example.full_description.to_s,
+            'described_class' => example.metadata[:described_class].to_s
+          }
+          span = tracer.start_span(example.description, attributes: attributes, start_timestamp: current_timestamp)
+          track_span(span)
+        end
+
+        def example_finished(notification)
+          pop_and_finalize_span do |span|
+            result = notification.example.execution_result
+            notification.example.metadata
+
+            span.set_attribute('result', result.status.to_s)
```

Comment:
I think one of the key value-adds of test instrumentation is the ability to track failing/flaky tests. In OpenTelemetry I _think_ (we're kinda invented spec as we go here) span.status is the appropriate place to record that information. I'd like to see span.status get set here if possible if we _know_ it's a failed test, `passed` / `failed` / `skipped` appear to be the possible values (I actually am rustier with rspec than i'd like to admit so entirely possible i'm wrong), so i think in the case of `result.status == :failed` we should set the span status to error. This is the approach some OSS vendor approaches take, like our good friends at [ddog](https://github.com/DataDog/dd-trace-rb/blob/988431adf07bb09e41bb845d4536816b0e61579a/lib/datadog/ci/test.rb#L58-L71), and also an approach taken by other OpenTelemetry Test instrumentations floating around in the community, like @rakyll's work in go-test-trace: https://github.com/rakyll/go-test-trace/blob/732baad7c542b6afa16c38eca52c34172a619b5e/main.go#L149-L151

An example of how to use the `Trace::Status` helpers can be found here: https://github.com/open-telemetry/opentelemetry-ruby/blob/87e0ccb0e30bf179ee826d67020ba600216afd29/instrumentation/redis/lib/opentelemetry/instrumentation/redis/patches/client.rb#L49. Wdyt? Does this make sense for your use case specifically?

Reply:
Yes, that works for me.